src/libotutil/ot-variant-builder.c

```diff
@@ -711,6 +737,300 @@ gvs_calculate_total_size (gsize body_size,
+        return glnx_throw_errno_prefix (error, "pread");
+      if (bytes_read == 0)
+        return glnx_throw (error, "Unexpected EOF in variant");
+      dst += bytes_read;
+      size -= bytes_read;
+      offset += bytes_read;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+gvs_fd_read_unaligned_le (int      fd,
+                          gint64   offset,
+                          gsize    size,
+                          guint64 *res,
+                          GError **error)
+{
+  union
+  {
+    guchar bytes[8];
+    guint64 integer;
+  } tmpvalue;
+
+  tmpvalue.integer = 0;
+
+  if (!gvs_fd_read_all (fd, offset, size, tmpvalue.bytes, error))
+    return FALSE;
+
+  *res = GUINT64_FROM_LE (tmpvalue.integer);
+  return TRUE;
+}
+
+/* It would be more natural for this to be in ot-variant-utils.c, but it depends on the internal
+ * copy of the glib variant type info code above so we put it here. */
+gboolean
+ot_variant_fd_get_child (int                  fd,
+                         gint64               offset,
+                         guint64              size,
+                         const GVariantType  *type,
+                         gsize                index,
+                         gint64              *child_offset,
+                         guint64             *child_size,
+                         const GVariantType **child_type,
+                         GError             **error)
+{
+  GVariantTypeInfo *type_info = g_variant_type_info_get (type);
+  GVariantTypeInfo *child_type_info = NULL;
+  const GVariantMemberInfo *member_info;
+  ArrayInfo *array_info;
+  gsize offset_size;
+  guint64 start, end, last_end;
+
+  switch (g_variant_type_info_get_type_char (type_info))
+    {
+    case G_VARIANT_TYPE_INFO_CHAR_TUPLE:
+    case G_VARIANT_TYPE_INFO_CHAR_DICT_ENTRY:
+      member_info = g_variant_type_info_member_info (type_info, index);
+
+      if (member_info == NULL)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
+                       "Variant index %ld out of bounds", index);
+          return FALSE;
+        }
+
+      offset_size = gvs_get_offset_size (size);
+
+      /* Out of bounds check the offset table reads */
+      if (member_info->ending_type == G_VARIANT_MEMBER_ENDING_OFFSET)
+        {
+          if (offset_size * (member_info->i + 2) > size)
+            return glnx_throw (error, "Invalid gvariant");
+        }
+      else
+        {
+          if (offset_size * (member_info->i + 1) > size)
+            return glnx_throw (error, "Invalid gvariant");
+        }
+
+      if (member_info->i + 1)
+        {
+          if (!gvs_fd_read_unaligned_le (fd,
+                                         offset + size - offset_size * (member_info->i + 1),
+                                         offset_size,
+                                         &start, error))
+          return FALSE;
+        }
+      else
+        start = 0;
+
+      start += member_info->a;
```

Comment:
I think what would help me a lot is if you could annotate some of these functions with the glib functions they came from.  For example, these bits clearly come from `gvs_tuple_get_child()`.  So maybe a comment at the top like:

```
This code derives from:
  - https://gitlab.gnome.org/GNOME/glib/blob/1c177ce0ab36991d33f4fbe0cfe6e003ca121b7e/glib/gvariant-serialiser.c#L851
```